go/internal/services/caches/caches.go

```diff
@@ -90,18 +105,70 @@ func New(config Config) (Caches, error) {
 	}
 
 	apiById, err := cache.New(cache.Config[string, db.Api]{
-		Fresh:   10 * time.Second,
-		Stale:   24 * time.Hour,
-		Logger:  config.Logger,
-		MaxSize: 1_000_000,
-
+		Fresh:    10 * time.Second,
+		Stale:    24 * time.Hour,
+		Logger:   config.Logger,
+		MaxSize:  1_000_000,
 		Resource: "api_by_id",
 		Clock:    config.Clock,
 	})
 	if err != nil {
 		return Caches{}, err
 	}
 
+	if config.CacheInvalidationTopic != nil {
+		// Create cluster caches that automatically handle distributed invalidation
+		ratelimitNamespace, err = clustering.New(clustering.Config[db.FindRatelimitNamespace]{
+			LocalCache: ratelimitNamespace,
+			Topic:      config.CacheInvalidationTopic,
+			Logger:     config.Logger,
+			NodeID:     config.NodeID,
+		})
+		if err != nil {
+			return Caches{}, err
+		}
+
+		verificationKeyByHash, err = clustering.New(clustering.Config[db.FindKeyForVerificationRow]{
+			LocalCache: verificationKeyByHash,
+			Topic:      config.CacheInvalidationTopic,
+			Logger:     config.Logger,
+			NodeID:     config.NodeID,
+		})
+		if err != nil {
+			return Caches{}, err
+		}
+
+		apiById, err = clustering.New(clustering.Config[db.Api]{
+			LocalCache: apiById,
+			Topic:      config.CacheInvalidationTopic,
+			Logger:     config.Logger,
+			NodeID:     config.NodeID,
+		})
+		if err != nil {
+			return Caches{}, err
+		}
+
+		// Start consuming invalidation events
+		consumer := config.CacheInvalidationTopic.NewConsumer()
+		consumer.Consume(context.Background(), func(ctx context.Context, event *cachev1.CacheInvalidationEvent) error {
+			// Don't process our own events to avoid loops
+			if event.SourceInstance == config.NodeID {
+				return nil
+			}
+
+			// Route invalidation events to the appropriate cache
+			switch event.CacheName {
+			case ratelimitNamespace.Name():
+				ratelimitNamespace.Remove(ctx, event.CacheKey)
+			case verificationKeyByHash.Name():
+				verificationKeyByHash.Remove(ctx, event.CacheKey)
+			case apiById.Name():
+				apiById.Remove(ctx, event.CacheKey)
```

Comment:
this currently results in an infinte loop because each other node gets this topic and re-broadcasts it